### PR TITLE
bpo-41905: added abc.update_abstractmethods

### PR DESCRIPTION
**Pull request by bentheiii**
relevant python-ideas thread: https://mail.python.org/archives/list/python-ideas@python.org/thread/6BNJ3YSEBPHEPGXSAZGBW3TJ64ZGZIHE/

https://bugs.python.org/issue41905

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept this contribution by verifying everyone involved has signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

## Recognized GitHub username

We couldn't find a [bugs.python.org](https://bugs.python.org/) (b.p.o) account corresponding to the following GitHub usernames:

@bentheiii

This might be simply due to a missing "GitHub Name" entry in one's b.p.o account settings. This is necessary for legal reasons before we can look at this contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.




You can [check yourself](https://check-python-cla.herokuapp.com/) to see if the CLA has been received.

Thanks again for the contribution, we look forward to reviewing it!

File: Doc/library/abc.rst
Comment (ericvsmith) [inline Doc/library/abc.rst]: Should be "within a class decorator" (add the word "a").

File: Lib/abc.py
Comment (ericvsmith) [inline Lib/abc.py]: I'm not sure what this is trying to say: "will not be considered to require that method implemented". Could you reword it?

File: Lib/abc.py
Comment (ericvsmith) [inline Lib/abc.py]: I think "that implement methods" should be "that add abstract methods to the class".

File: Lib/abc.py
Comment (ericvsmith) [inline Lib/abc.py]: Typo in "subclassing".

File: Lib/dataclasses.py
Comment (ericvsmith) [inline Lib/dataclasses.py]: I know this is nitpicking, but all of the other comments in this file start with a capital letter and end in a period.

**Review by ericvsmith — DISMISSED** [review]
Although I'm no expert on ABC's, this looks good to me, with a few noted tiny changes. I'll do some more research before signing off on this, though.

My only real concern is that dataclasses now imports abc. We've jumped through hoops to make dataclasses not import typing, because it's slow to import. I don't know if abc brings the same issues, but I'll try to do some investigating. The trick it uses for typing is to check if 'typing' is in sys.modules before checking for typing.<somthing_or_other>. I'm not sure such trickery is warranted here. Probably not, since just starting the REPL looks like it imports abc (but not typing).

**Comment by bentheiii** [comment]
I have made the requested changes; please review again

File: Lib/test/test_abc.py
Comment (iritkatriel) [inline Lib/test/test_abc.py:542]: I think you need this test case to cover the first loop:

```
import abc
class FooABC(metaclass=abc.ABCMeta):
    @abc.abstractmethod
    def bar(self):
        pass

del FooABC.bar
assert ('bar' in FooABC.__abstractmethods__)
assert ('bar' not in FooABC.__dict__)

```

**Comment by ericvsmith** [comment]
Since `update_abstractmethods` is documented as being usable as a decorator, there should be a test for that. I realize it's trivially obvious that it works, but still. Although I guess you'll need another decorator that adds an abstract method (like dataclasses used to do) in order to test it. And then use them together.

File: Lib/abc.py
Comment (iritkatriel) [inline Lib/abc.py]: Since you didn't check in this function that it's an ABC, it's not guaranteed that `cls.__abstractmethods__` works:

```
>>> class FooABC: pass
... 
>>> FooABC.__abstractmethods__
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: __abstractmethods__

```
Maybe add a test for when this function is called for a non-ABC.

**Comment by ericvsmith** [comment]
> 
> 
> Since `update_abstractmethods` is documented as being usable as a decorator, there should be a test for that. I realize it's trivially obvious that it works, but still. Although I guess you'll need another decorator that adds an abstract method (like dataclasses used to do) in order to test it. And then use them together.

Change looks good. Thanks.

**Comment by bentheiii** [comment]
> My only real concern is that dataclasses now imports abc. We've jumped through hoops to make dataclasses not import typing, because it's slow to import. I don't know if abc brings the same issues, but I'll try to do some investigating. The trick it uses for typing is to check if 'typing' is in sys.modules before checking for typing.<somthing_or_other>. I'm not sure such trickery is warranted here. Probably not, since just starting the REPL looks like it imports abc (but not typing).

If this does prove to be an issue, then we can change the `isinstance(cls, ABCMeta)` checking to `hasattr(cls, '__abstractmethods__')` (no sys.modules check necessary), and then import `abc.update_abstractmethods` before calling it inside the "if" block.

File: Lib/functools.py
Comment (rhettinger) [inline Lib/functools.py]: The nested *getattr()* isn't very clear.  

Also, since this is in the inner loop for a dynamic decorator, it will slow the class creation process.  I didn't used to think that was important but experience with named tuples show that people care a great deal about the amount of computation that occurs during an import.

Comment (bentheiii) [reply Lib/functools.py]: would using a walrus operator be okay here?

```python
roots = {...
         if (root:=getattr(cls, op, None)) is not getattr(object, op, None)
         and not getattr(root, '__isabstractmethod__', False)}
```

Comment (ericvsmith) [reply Lib/functools.py]: That seems fine to me.

**Comment by ericvsmith** [comment]
[On importing abc into dataclasses]
> 
> If this does prove to be an issue, then we can change the `isinstance(cls, ABCMeta)` checking to `hasattr(cls, '__abstractmethods__')` (no sys.modules check necessary), and then import `abc.update_abstractmethods` before calling it inside the "if" block.

I don't think this will be an issue, since abc seems to always be imported on startup, anyway.

File: Lib/abc.py
Comment (iritkatriel) [inline Lib/abc.py:154]: Add tests for deeper inheritance - I think iterating over `__bases__` may not be enough:

```
>>> class A: pass
... 
>>> class B(A): pass
... 
>>> class C(B): pass
... 
>>> C.__bases__
(<class '__main__.B'>,)
>>> B.__bases__
(<class '__main__.A'>,)
```

Comment (bentheiii) [reply Lib/abc.py:154]: It actually is- we are assured that the parent's `__abstractmethods__` are correct since, after a class is subclassed, its abstraction status cannot change:
```
class A(ABC):
  @abstractmethod
  def foo(self): pass

class B(A):
  pass

# after this point, A.__abstractmethods__ will never change, so we can rely on it always being correct when checking B's Abstraction status
```

Comment (bentheiii) [reply Lib/abc.py:154]: I added a test case regardless

Comment (iritkatriel) [reply Lib/abc.py:154]: I see, interesting. How about this case? I think you might need to break after the first hit for the method:

```
class A(metaclass=abc.ABCMeta):
    @abc.abstractmethod
    def foo(self):
        pass

class B(metaclass=abc.ABCMeta):
    @abc.abstractmethod
    def foo(self):
        pass

class C(A, B):
    pass

A.foo = lambda self: None


assert(C.foo is not abstract)
```

Comment (bentheiii) [reply Lib/abc.py:154]: This is expected (albeit confusing) behavior. Since abstraction status cannot change after subclassing, A is still considered abstract, even though its method is implemented. However, since namespace resolution occurs through the `__dict__` attribute, The method is defined as far as C is concerned. Indeed, if you were to call `update_abstractmethods(C)` after this code, C would become concrete and usable, since its `foo` method implemented.

In essence, abstract methods shouldn't be changed after the class is used, so this use case is not handled by this implementation.

Comment (iritkatriel) [reply Lib/abc.py:154]: I see. I was concerned about something a bit different, but the test I suggested didn't quite capture that.

See the code below. I wondered if the last update_abstractmethods will know to ignore A and look only at B. It works because you put 'value = getattr(cls, name, None)'  (cls and not scls). So I guess it checks this once for A and once for B, but gets the same (correct) result both times.

Might be worth adding this test to make sure nobody breaks that getattr. 

```
import abc

class A(metaclass=abc.ABCMeta):
    @abc.abstractmethod
    def foo(self):
        pass

class B(metaclass=abc.ABCMeta):
    def foo(self):
        pass

class C(B, A):
    @abc.abstractmethod
    def foo(self):
        pass

abc.update_abstractmethods(C)

del C.foo

abc.update_abstractmethods(C)
```

File: Lib/abc.py
Comment (gvanrossum) [inline Lib/abc.py]: I would like the summary to fit on one line.
```suggestion
    """Recalculate the set of abstract methods of an abstract class.
```

File: Lib/abc.py
Comment (gvanrossum) [inline Lib/abc.py]: I'm not sure I like this. This function is potentially expensive. It could even be detrimental -- I could imagine some very dynamic code that's messing around dynamically with a base class and then calls `update_abstractmethods()` for all subclasses. So I propose to remove this check (and the documentation about it).

File: Lib/abc.py
Comment (gvanrossum) [reply Lib/abc.py:154]: Well, I disagree with the check for empty `__subclasses__()`, and I think in Irit's last example the right thing to do is to call `abc.update_abstractmethods(A)` and then `abc.update_abstractmethods(C)`.

Note that it's not technically true that the abstractness cannot change once subclassed -- it may be undocumented but one could just write `A.__abstractmethods__ = ()`.

FWIW I do agree that the code should *assume* that the abstractness of the bases is correct. If some base lies about it, well, Garbage In, Garbage Out.

File: Lib/test/test_abc.py
Comment (gvanrossum) [reply Lib/test/test_abc.py:542]: @iritkatriel What does that test case accomplish other than showing that FooABC is now in an inconsistent state?

File: Lib/abc.py
Comment (bentheiii) [reply Lib/abc.py]: That might be okay, However, removing this check means that we have to call `update_abstractmethods` on all of `cls`'s subclasses, meaning that `cls.__subclasses__()` will still be called, so performance will not improve.

File: Lib/test/test_abc.py
Comment (iritkatriel) [reply Lib/test/test_abc.py:542]: @gvanrossum In this comment I left out the update_abstractmethods, see Ben's version which he committed as test_update_del.  Without this test the first loop in update_abstractmethods was not exercised.

File: Lib/abc.py
Comment (iritkatriel) [reply Lib/abc.py:154]: I don't think abc.update_abstractmethods(A) would have any impact here, A never changed.

Comment (gvanrossum) [reply Lib/abc.py:154]: Sorry, I was commenting on your earlier example which has `A.foo = lambda ...` -- in that example it certainly would have an effect. :-)

My disagreement was with this quote from @bentheiii:

> In essence, abstract methods shouldn't be changed after the class is used, so this use case is not handled by this implementation.

I agree with @iritkatriel that adding that test would be useful (I didn't verify -- maybe it's already added?).

File: Lib/abc.py
Comment (gvanrossum) [reply Lib/abc.py]: Only if you're not sure that there are no subclasses. In `@dataclass` we can assume there are no subclasses, and that's so in most class decorators (which are only applicable to freshly created classes).

I recommend that `update_abstractmethods()` should *not* try to fix up subclasses -- it's just the case that an ambitious dynamic class hacker could take two documented features (this and `__subclasses__()`) and create something that automatically updates affected subclasses.

File: Lib/abc.py
Comment (iritkatriel) [reply Lib/abc.py:154]: > Sorry, I was commenting on your earlier example which has `A.foo = lambda ...` -- in that example it certainly would have an effect. :-)
> 

Ah, in that case it raises an exception that A has been subclasses so you can't update it anymore. I think there's a test for that already but if not then there should be.

File: Lib/abc.py
Comment (bentheiii) [reply Lib/abc.py]: I think that could cause a problem if user chain two decorators: one that updates subclasses and one that doesn't

```python

def decoratorA(cls):
  """implements methods and creates a subclass, calling
  update_abstractmethods on cls and its subclasses
  """
  ...

def decoratorB(cls):
  """implements methods only, not updating subclasses
  """
  ...

@decoratorB
@decoratorA
class A(SomeABC):
  ...

# A's subclass is now out of sync with decoratorB's changes
```

Thinking about it, the previous implementation was indeed faulty since, even if a decorator that needed to subclass `cls` was responsible enough to delay subclassing until after calling `update_abstractmethods`, that would meant that follow-up decorators simply couldn't call it.

The only sensible solution is to call update_abstractmethods on all of the class's subclasses with each call (`@dataclass` can't be certain that the subject class wasn't subclassed in a previous mixin decorator).

File: Lib/abc.py
Comment (iritkatriel) [reply Lib/abc.py:154]: I see now - you're suggesting to remove that exception. In that case this test is probably needed.

File: Lib/abc.py
Comment (gvanrossum) [reply Lib/abc.py]: If you do that, you get what you deserve.

Please keep it simple.

Comment (bentheiii) [reply Lib/abc.py]: It's implemented in the latest commit, but I must say I'm not sure what you mean. Who "gets what they deserve"?
* Whoever implemented `decoratorA` (which may well be `@dataclass`) was right to assume there are no subclasses, since it did not create them, and wanted to avoid unnecessary checks.
* Whoever implemented `decoratorB` did their due diligence by making sure it returned the class and its subclasses in a synced state.
* Whoever wrote `class A` may be totally unaware that `decoratorB` creates subclasses or that `decoratorA` doesn't take them into account. Even if they did,  they would need to implement a fix themselves, which would require at the least to iterate through `A.__subclasses__()`. Even that won't keep them 100% safe, however, since these subclasses can have subclasses of their own. Very far fetched, perhaps, but that's exactly what very dynamic code might produce.

Comment (bentheiii) [reply Lib/abc.py]: Truth be told, if we do assume that decorators and tools are allowed to subclass (and perhaps other permanent operations like instancing) before calling `update_abstractmethods`, I'm leaning closer and closer to an automatic solution as proposed by Bar and Reymond.

File: Doc/library/abc.rst
Comment (gvanrossum) [inline Doc/library/abc.rst]: I prefer to do less editorializing.

```suggestion
   .. note::

      This function assumes that *cls*'s superclasses are already updated.
      It does not update any subclasses.
```

File: Lib/abc.py
Comment (gvanrossum) [inline Lib/abc.py:125]:
```suggestion

def update_abstractmethods(cls):
```

File: Lib/abc.py
Comment (gvanrossum) [inline Lib/abc.py:160]:
```suggestion
    return cls

```

**Review by gvanrossum — COMMENTED** [review]
I just don't care enough about the purely hypothetical scenario you're describing to want to complicate this function. The automatic solution also has its problems (to begin, it would require C code).

**Comment by gvanrossum** [comment]
Just noted this:

@ericvsmith
> My only real concern is that dataclasses now imports abc.

I looked into this, dataclasses already imports functools which already imports abc, so there's no new module being loaded. Also, abc.py is pretty compact and backed by C code -- most of the Python code is in `_py_abc.py` which is only loaded if `_abc` (the C code) cannot be loaded for some reason.

So I don't think there's a worry. @ericvsmith, all your other comments seem to have been taken care of this -- can you sign off? (Unless you want to jump into the debate about subclasses.)

**Comment by ericvsmith** [comment]
@gvanrossum: Agreed about importing abc. I'll let you and the others work out the subclass issue.

**Comment by bentheiii** [comment]
Of course, I defer judgement to the dev team, but I'd be remiss if I didn't reflect the implications of this behavior:

When this behavior is implemented, it will effectively split all class decorators into one of three categories:

1. **ABC-agnostic**, who don't consider abstract superclasses as a use case. `total_ordering` will be of this category.
2. **ABC-aware, subclass-agnostic**, who accept the possibility of an abstract superclass, and thus call `update_abstractmethods`. `dataclass` will be of this of this category, as well as, `update_abstractmethods`, techinally.
3. **ABC-aware, subclass-aware**, who accept the possibility of an abstract superclass, and call `update_abstractmethods`, and also accept possibility of it having subclasses, so they also call `update_abstractmethods` on its subclasses (perhaps even recursively).

An important consequence is, that decorators can only be reliably stacked on top of each other *only* if they are of the same category. And while `update_abstractmethods` can be used to adapt a category 1 decorator to category 2, there are no simple means to adapt to category 3.

Of course, the mere existence of a decorator of the third category is a case so far fetched that I doubt it even exists across all python code. This is a valid reason to discount it, but the implication is that if were to ever exist, it could not be used with virtually any other decorator.

**Review by gvanrossum — APPROVED** [review]
You can consider us amply warned.

**Comment by gvanrossum** [comment]
@rhettinger any final word? @iritkatriel any comments?

**Comment by iritkatriel** [comment]
> @rhettinger any final word? @iritkatriel any comments?

A very pedantic one, which should not block this PR being merged.  The doc says "It does not update any subclasses", but no test asserts this.

**Review by bharel — APPROVED** [review]

**Comment by gvanrossum** [comment]
Thanks! All merged.

**Pull request merged**
